lxd/firewall/drivers/drivers_util.go

```diff
@@ -36,3 +36,51 @@ func portRangeStr(portRange [2]uint64, delimiter string) string {
 	}
 	return fmt.Sprintf("%d%s%d", portRange[0], delimiter, portRange[0]+portRange[1]-1)
 }
+
+// getOptimisedDNATRanges returns a map of listen port ranges to target port ranges that can be
+// applied in any order.
+func getOptimisedDNATRanges(forward *AddressForward) map[[2]uint64][2]uint64 {
+	targetPortsLen := len(forward.TargetPorts)
+	listenPortsLen := len(forward.ListenPorts)
+
+	listenPortRanges := PortRangesFromSlice(forward.ListenPorts)
+	targetPortRanges := PortRangesFromSlice(forward.TargetPorts)
+
+	snatRules := make(map[[2]uint64][2]uint64, listenPortsLen)
+
+	targetPortRangeMap := make(map[uint64]uint64, len(targetPortRanges))
+	for _, targetPortRange := range targetPortRanges {
+		targetPortRangeMap[targetPortRange[0]] = targetPortRange[1]
+	}
+
+	nProcessedPorts := 0
+	for _, listenPortRange := range listenPortRanges {
+		rangeEndIdx := nProcessedPorts + int(listenPortRange[1])
+
+		if targetPortsLen == 1 {
+			// Only one target port. Each listen range maps directly to that port.
+			snatRules[listenPortRange] = targetPortRanges[0]
+			nProcessedPorts += int(listenPortRange[1])
+
+		} else if targetRangeSize, ok := targetPortRangeMap[forward.TargetPorts[nProcessedPorts]]; ok && forward.TargetPorts[nProcessedPorts] == listenPortRange[0] {
```

Comment:
We try and avoid using this form of if statement, (i.e initialising a local variable inside the if statement itself).
The reason for this is that it complicates the scope, and in this case, makes for a harder to read longer line that goes out of the github viewport.